### PR TITLE
Fix #10899: Restore None hyperlink in py domain

### LINKED ISSUE #1152: `None` type does not get linked in Sphinx
There is a bug in Sphinx where the `None` type does not get linked.

https://github.com/sphinx-doc/sphinx/issues/10899.

I wasted some time looking for why this was the case. There's nothing to be done here in icalendar, but it would be great to contribute a PR upstream in Sphinx to restore this feature. Then we can have it in icalendar.

_Originally posted by @stevepiercy in [#1115](https://github.com/collective/icalendar/pull/1115/changes#r2778311030)_

### LINKED ISSUE #10899: py domain: `None` is no longer hyperlinked after 4.4.0
### Describe the bug

Since Sphinx 4.4.0, `None` is no longer hyperlinked.

I think this might be in all cases, but I've noticed it in the case where it's used as one of the types for a parameter through a docstring, using autodoc and napoleon (and intersphinx).

Sphinx 4.4.0 is the first version this started happening in, and it's persisted in every version since.

I think I've been able to narrow it down as due to https://github.com/sphinx-doc/sphinx/commit/5da68c3de3a05f0498b1576f01397be6bad82f2d (#10021).

### How to Reproduce

With Sphinx 4.3.2:
```
$ git clone https://github.com/tweepy/tweepy
$ cd tweepy
$ pip install -r requirements.txt
$ cd docs
$ make html
$ # open _build/html/client.html and see None properly hyperlinked
```

With Sphinx 4.4.0:
```
$ git clone https://github.com/tweepy/tweepy
$ cd tweepy
$ pip install -r requirements.txt
$ cd docs
$ make html
$ # open _build/html/client.html and see None not hyperlinked
```

### Your project

https://github.com/tweepy/tweepy

### OS

Microsoft Windows [Version 10.0.19043.1889]

### Python version

Python 3.10.0 (tags/v3.10.0:b494f59, Oct  4 2021, 19:00:18) [MSC v.1929 64 bit (AMD64)] on win32

### Sphinx version

4.4.0 and newer

### Sphinx extensions

hoverxref.extension
myst_parser
sphinx.ext.autodoc
sphinx.ext.autosectionlabel
sphinx.ext.intersphinx
sphinx.ext.napoleon
sphinx_search.extension
sphinx_tabs.tabs

### PR DESCRIPTION
**Pull request by japinderofficial-hub**
## Description

Since Sphinx 4.4.0, the `None` singleton is no longer hyperlinked in type annotations.

This occurs because `None` is currently treated using the `class` role instead of the `obj` role during cross-reference generation.

However, `None` is not a type — it is a singleton object. Therefore, it should be referenced using the `:obj:` role to generate correct cross-references.

## Changes

- Override the `make_xref()` method in `PyField` and `PyTypedField`
- Detect when `rolename == "class"` and `target == "None"`
- Switch the role from `"class"` to `"obj"` to ensure proper linking
- Add test coverage to verify correct cross-referencing of `None`

## Related Issues

Closes #10899  
Closes collective/icalendar#1152

### CODE REVIEW

**Comment by niccokunzmann** [comment]
@japinderofficial-hub From my view, the comments by copilot are valid - especially, I do not like code duplication as a maintainer. The edge cases need considering and also testing - I would value it if you could add tests for them if I was the maintainer.
When the tests pass, you can let us know over at icalendar and we can give it a go, build the documentation and see if the changes address the problem for us.

File: sphinx/domains/python/_object.py [hunk 1]
```diff
@@ -151,15 +151,49 @@ def make_xrefs(
...
+        rolename: str,
+        domain: str,
+        target: str,
+        innernode: type[TextlikeNode] = nodes.emphasis,
+        contnode: Node | None = None,
+        env: BuildEnvironment | None = None,
+        inliner: Inliner | None = None,
+        location: Node | None = None,
+    ) -> Node:
+        if rolename == 'class' and target == 'None':
```
Comment (niccokunzmann) [inline sphinx/domains/python/_object.py:190]: Personally, I would move the if check into is_none(rolename, target) and test that independently and remove the duplication.

Same goes for the `rolename = 'obj'`

Something like `rolename = self.choose_rolename(...)`

That can be documented, then.

File: tests/test_domains/test_domain_py_fields.py [hunk 2]
```diff
@@ -609,3 +609,31 @@ def test_type_field(app):
         reftarget='typing.Any',
         refspecific=False,
     )
+
+
+@pytest.mark.sphinx('html', testroot='_blank')
+def test_none_type_xref(app):
+    """Test that None type is properly linked with the obj role."""
+    text = (
```
Comment (niccokunzmann) [inline tests/test_domains/test_domain_py_fields.py:617]: Since you use pytest, please use pytest.mark.parametrize
to generate a separate test case for each special case. That makes it much easier for me to understand if they are all tested properly.

https://docs.pytest.org/en/stable/example/parametrize.html#different-options-for-test-ids

Especially

```
assert len(none_xrefs) >= 1,
```


does not tell me wether you really test all cases.
With parametrization, it should be changed to 
```
assert len(none_xrefs) == 1
```

I would say that the test then actually catches what it should. One xref per example.

**Review by niccokunzmann — CHANGES_REQUESTED** [review]
Hi, I left some comments on the code :)
I am not familiar with sphinx but with pytest and code quality.
